lib/deref.js

```diff
@@ -247,23 +247,29 @@ module.exports = {
       }
       return { value: 'reference ' + schema.$ref + ' not found in the OpenAPI spec' };
     }
-    if (concreteUtils.compareTypes(schema.type, SCHEMA_TYPES.object) || schema.hasOwnProperty('properties') ||
-      schema.hasOwnProperty('additionalProperties')) {
+
+    const getAdditionalProperties = (schema) => {
+      let additionalProperties;
+      if (_.isBoolean(schema.additionalProperties)) {
+        additionalProperties = schema.additionalProperties;
+      }
+      else {
+        additionalProperties = this.resolveRefs(schema.additionalProperties, parameterSourceOption,
+          components, schemaResolutionCache, resolveFor, resolveTo, stack, _.cloneDeep(seenRef), stackLimit);
+      }
+      return additionalProperties;
+    };
+
+    if (concreteUtils.compareTypes(schema.type, SCHEMA_TYPES.object) || schema.hasOwnProperty('properties')) {
```

Comment:
@erfemega Any reason for removing check that also checked if only `additionalProperties` is present (i.e. `type` and `properties` are not defined), then we should be considering this if block?

This change will make so following kind of schemas not being resolved correctly anymore. I've confirmed this as well by testing with this branch that in such case we're having empty object being generated.

```
additionalProperties:
  type: string
```

Reply:
Hi @VShingala. You were right, I was skipping those scenarios.
Now it's fixed.
I've applied the following changes to resolve it:
- Returning the condition when schema only contains additionalProperties. It will be resolved as object type.
- The condition was modified to resolve the schema as object type only if it doesn't have type or if type is object
- If the schema is typed as array, it will be resolved as array type and, as additionalProperties is not defined for array type, it will ignore the additionalProperties (it was validated in swagger editor).
- If schema have additionalProperties and any other type different than object and array  it will ignore the additionalProperties (it was validated with swagger editor)
- Added some scenarios to validate that 'additionalProperties' is being resolved correctly.


Please let me know your comments.